provider/core/src/refcount.rs

```diff
@@ -0,0 +1,144 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+use core::mem::size_of;
+use core::{mem::ManuallyDrop, ptr::NonNull};
+use alloc::sync::Arc;
+use alloc::{string::String, rc::Rc};
+use alloc::boxed::Box;
+use stable_deref_trait::{StableDeref, CloneStableDeref};
+
+type WrappedType = Option<Box<[u8]>>;
+
+// Safe as a sentinel because 0x1 is not a valid usize address
+const SENTINEL: *const WrappedType = 1usize as *const WrappedType;
```

Comment:
nit: make this a generic const fn that returns `*const T`

Reply:
Hmm, this is a good suggestion when we make it generic. For now I have no generics which makes the safety easier to reason about.